one-offs/images-from-boms.go

```diff
@@ -0,0 +1,187 @@
+//	GitPrefix          string `yaml:"gitPrefix"`
+//	GoogleImageProject string `yaml:"gitPrefix,omitempty"`
+//}
+
+type Image struct {
+	Digest string   `json:"digest"`
+	Tags   []string `json:"tags"`
+}
+
+func main() {
+	flag.Parse()
+
+	ctx := context.Background()
+	storageSvc, err := storage.NewClient(ctx, option.WithCredentialsFile(*jsonKeyPath), option.WithScopes(storage.ScopeReadOnly))
+	if err != nil {
+		log.Fatalf("Error generating new storage client: %v", err)
+	}
+
+	boms := make([]*Bom, 0, 100)
+	iter := storageSvc.Bucket(*bucket).Objects(ctx, &storage.Query{Prefix: "bom/"})
+	for obj, err := iter.Next(); err == nil; obj, err = iter.Next() {
+		if releaseRegexp.Match([]byte(obj.Name)) {
+			log.Println(obj.Name)
+			r, err := storageSvc.Bucket(*bucket).Object(obj.Name).NewReader(ctx)
+			if err != nil {
+				log.Fatalf("error reading object %v: %v", obj.Name, err)
+			}
+			bom := &Bom{}
+			if err = yaml.NewDecoder(r).Decode(bom); err != nil {
+				log.Fatalf("error decoding bom %v: %v", obj.Name, err)
+			}
+			boms = append(boms, bom)
+			log.Printf("%+v", bom)
+			//break
+		}
+	}
+
+	// Go get all the images and index them by tag
+	imagesByTag := make(map[string]*Image, 1000)
+	l := boms[0].Services.List()
+	for svcElem := l.Front(); svcElem != nil; svcElem = svcElem.Next() {
+		svc := svcElem.Value.(*Service)
+		log.Printf("Executing gcloud for %v\n", svc.name)
+		cmd := exec.Command("gcloud", "container", "images", "list-tags", fmt.Sprintf("gcr.io/%v/%v", *fromProjID, svc.name), "--format", "json")
+		stdout, err := cmd.StdoutPipe()
+		if err != nil {
+			log.Fatalf("error getting gcloud stdout: %v", err)
+		}
+		if err := cmd.Start(); err != nil {
+			log.Fatalf("error starting: %v", err)
+		}
+
+		images := make([]*Image, 0, 100)
+		if err := json.NewDecoder(stdout).Decode(&images); err != nil {
+			log.Fatalf("error decoding: %v", err)
+		}
+		if err := cmd.Wait(); err != nil {
+			log.Fatalf("error waiting: %v", err)
+		}
+
+		count := 0
+		for _, img := range images {
+			for _, tag := range img.Tags {
+				fullTag := fmt.Sprintf("%v:%v", svc.name, tag)
+				imagesByTag[fullTag] = img
+			}
+			count++
+		}
+		log.Printf("Indexed %v tags\n", count)
+		//break
+	}
+
+	b := strings.Builder{}
+	for _, bom := range boms {
+		l := bom.Services.List()
+		for svcElem := l.Front(); svcElem != nil; svcElem = svcElem.Next() {
+			svc := svcElem.Value.(*Service)
+			fullTag := fmt.Sprintf("%v:%v", svc.name, svc.Version)
+
+			img, ok := imagesByTag[fullTag]
+			if !ok {
+				log.Printf("Can't find %v for bom %v\n", fullTag, bom.Version)
+				continue
+			}
+
+			b.WriteString(fmt.Sprintf("gcloud container images add-tag gcr.io/%v/%v gcr.io/%v/%v:spinnaker-release-%v\n", *fromProjID, fullTag, *toProjId, svc.name, bom.Version))
+			for _, tag := range img.Tags {
+				b.WriteString(fmt.Sprintf("gcloud container images add-tag gcr.io/%v/%v gcr.io/%v/%v:%v\n", *fromProjID, fullTag, *toProjId, svc.name, tag))
```

Comment:
Won't this do a lot of duplicate work, or am I reading this wrong?  It looks like for each BOM we find we're writing out a command to copy all tags for each microservice (rather than just the ones relevant to that BOM).  Should this like be outside of the BOM loop (and the BOM loop just adds the `spinnaker-` tags)?  That being said, the output doesn't seem to have all these duplicates so maybe I'm reading it wrong.

Reply:
@plumpy mentioned that there are a bunch of extra tags now being added by the `spinrel` tool. Not mention that there are a ton of variants that aren't addressed here (looking at you, `-slim`, `-ubuntu`, `-java8`, etc).

So instead of hardcoding what tags are important, I just grabbed them all. 

That being said, I changed this to do everything in one command, since the format is `gcloud container image add-tag SRC DEST [DEST ...]. Now there are about ~1700 commands instead of 8k

Reply:
I definitely agree with grabbing the tags rather than hard-coding them...what I was worried about here is that if two BOMs both point to the same image (as often happens for patch releases) we're going to copy them over multiple times.  The original version of the script, for example, had this line 7 times because a bunch of 1.12.x patch releases had the same rsoco version:
```
gcloud container images add-tag gcr.io/spinnaker-marketplace/rosco:0.9.0-20190123170846 gcr.io/spinnaker-community/rosco:0.9.0-20190123170846
```

Reply:
It looks like there's still some duplication with the current file:
```
$ wc -l migrate-containers.sh
    3153 migrate-containers.sh
$ sort migrate-containers.sh  | uniq | wc -l
    2453
```
Maybe it doesn't really matter, or maybe we could just post-filter...

Reply:
Fixed.